### PR TITLE
Sharding sample ported to java and 2.6

### PR DESCRIPTION
**Pull request by johanandren**

### CODE REVIEW

**Comment by johanandren** [comment]
In parallell i've updated the scala sample, see #181

File: akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/KillrWeather.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    List<Integer> seedNodePorts =
+        ConfigFactory.load().getStringList("akka.cluster.seed-nodes")
+          .stream()
+          .map(AddressFromURIString::parse)
+          .map(addr -> (Integer) addr.port().get()) // Missing Java getter for port, fixed in Akka 2.6.2
+          .collect(Collectors.toList());
+
+    // Either use a single port provided by the user, or start the all listed seed nodes plus one node on a random port
+    // in this JVM if the user didn't provide args for the app
+    // In a production application you wouldn't typically start multiple ActorSystem instances in the
```
Comment (chbatey) [inline akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/KillrWeather.java]: -typically?

File: akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/KillrWeather.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    // same JVM, here we do it to easily demonstrate these ActorSystems (which would be in separate JVM's)
```
Comment (chbatey) [inline akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/KillrWeather.java]: -easily?

File: akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherStation.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Generally geo-based data, unless it is a floating station.
+ *
+ * This is just an example: cluster sharding would be overkill for just keeping a small amount of data,
+ * but becomes useful when you have a collection of 'heavy' actors (in terms of processing or state)
```
Comment (chbatey) [inline akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherStation.java]: Maybe remove the heavy section, other reasons for using sharding e.g. resiliency, regardless of how heavy they are

File: akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherRoutes.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    timeout = system.settings().config().getDuration("killrweather.routes.ask-timeout");
+    // use a pre-configured object mapper from akka-jackson also for HTTP JSON
+    // this lets us use the -parameters compiler argument to skip annotating field names on immutable classes
+    objectMapper = JacksonObjectMapperProvider.get(system).getOrCreate("jackson-json", Optional.empty());
+    dataUnmarshaller = Jackson.unmarshaller(objectMapper, WeatherStation.Data.class);
+  }
+
+  private CompletionStage<WeatherStation.DataRecorded> recordData(long wsid, WeatherStation.Data data) {
+    EntityRef<WeatherStation.Command> ref = sharding.entityRefFor(WeatherStation.TypeKey, Long.toString(wsid));
+    return ref.ask(replyTo -> new WeatherStation.Record(data, System.currentTimeMillis(), replyTo), timeout);
```
Comment (chbatey) [inline akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherRoutes.java]: Maybe point this line out in the readme as the magic of cluster sharding

File: akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherRoutes.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  private CompletionStage<WeatherStation.QueryResult> query(long wsid, WeatherStation.DataType dataType, WeatherStation.Function function) {
+    EntityRef<WeatherStation.Command> ref = sharding.entityRefFor(WeatherStation.TypeKey, Long.toString(wsid));
+    return ref.ask(replyTo -> new WeatherStation.Query(dataType, function, replyTo), timeout);
```
Comment (chbatey) [inline akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherRoutes.java]: Or this one

File: akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherStation.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+ */
+final class WeatherStation extends AbstractBehavior<WeatherStation.Command> {
+
+  // setup for using WeatherStations through Akka Cluster Sharding
+  // these could also live elsewhere and the WeatherStation class be completely
+  // oblivious to being used in sharding
+  public static final EntityTypeKey<Command> TypeKey =
+    EntityTypeKey.create(WeatherStation.Command.class, "WeatherStation");
+
+  public static void initSharding(ActorSystem<?> system) {
```
Comment (chbatey) [inline akka-sample-sharding-java/killrweather/src/main/java/sample/killrweather/WeatherStation.java]: Maybe point this out in the readme, and that this is a common pattern for sharded actors

File: akka-sample-sharding-java/README.md [hunk 7]
```diff
@@ -1,39 +1,164 @@
-This tutorial contains a sample illustrating [Akka Cluster Sharding](http://doc.akka.io/docs/akka/current/java/cluster-sharding.html#an-example).
+# Cluster Sharding sample
 
-## Example overview
+The KillrWeather sample illustrates how to use [Akka Cluster Sharding](http://doc.akka.io/docs/akka/current/java/typed/cluster-sharding.html).
+It also shows the basic usage of [Akka HTTP](https://doc.akka.io/docs/akka-http/current/index.html).
+
```
Comment (chbatey) [inline akka-sample-sharding-java/README.md]: Maybe a short intro of the sceanrio before jumping into code references?

**Pull request merged**